ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +193,452 @@ let order_f (_, vm_rec) =
+               which can run the vm: (%s %s)."
+              (Ref.string_of h)
+              (Db.Host.get_name_label ~__context ~self:h) ;
+
+            vm_anti_affinity_spread_evenly_plan ~__context remaining_vms
+              vm_can_boot_on_host
+              (grp_ranked_hosts
+              |> update_anti_affinity_grp_ranked_hosts vm_size group h
+              )
+              ((vm, h) :: mapping)
+          with
+          | Api_errors.Server_error (error_code, []) as e
+          when error_code = Api_errors.no_hosts_available
+          ->
+            debug
+              "Spread evenly plan: no host can run this anti-affinity VM. Stop \
+               the planning as there won't be a valid plan for this VM." ;
+            raise e
+        )
+      | _ ->
+          debug
+            "Spread evenly plan: not an anti-affinity VM, skip it, continue \
+             with the next one." ;
+          vm_anti_affinity_spread_evenly_plan ~__context remaining_vms
+            vm_can_boot_on_host grp_ranked_hosts mapping
+    )
+
+(** Try to get a no breach plan for anti-affinity VMs (for each anti-affinity group, there are at least
+    2 hosts having running VMs in the group): 
+    1. For all the anti-affinity VMs which sorted in an increasing order in terms of the VM's memory
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, try to select a host for it so that there are at least 2 hosts which
+       has running VMs in the same anti-affinity group. If there are already 2 hosts having running VMs
+       in that group, skip planning for the VM. *)
+let rec vm_anti_affinity_no_breach_plan ~__context vms_increasing
+    vm_can_boot_on_host grp_ranked_hosts mapping =
+  debug "vm_anti_affinity_no_breach_plan" ;
+  log_vms_increasing ~__context vms_increasing ;
+  log_grp_ranked_hosts ~__context grp_ranked_hosts ;
+  match vms_increasing with
+  | [] ->
+      mapping
+  | (vm, vm_size) :: remaining_vms -> (
+    match Xapi_vm_helpers.vm_has_anti_affinity ~__context ~vm with
+    | Some (`AntiAffinity group) ->
+        debug "No breach plan: try to plan for VM (%s %s)." (Ref.string_of vm)
+          (Db.VM.get_name_label ~__context ~self:vm) ;
+        let ranked_hosts =
+          try VMGroupMap.find group grp_ranked_hosts
+          with Not_found ->
+            error "Failed to find ranked_hosts in grp_ranked_hosts for group." ;
+            raise (Api_errors.Server_error (Api_errors.internal_error, []))
+        in
+        let no_resident_hosts =
+          ranked_hosts |> IntMap.find_opt 0 |> Option.value ~default:[]
+        in
+        let resident_hosts_cnt =
+          ranked_hosts
+          |> IntMap.bindings
+          |> List.fold_left
+               (fun cnt (vm_cnt, hosts) ->
+                 match vm_cnt with 0 -> cnt | _ -> cnt + List.length hosts
+               )
+               0
+        in
+        if
+          List.length no_resident_hosts + resident_hosts_cnt > 1
```

Comment:
I think this could be implemented by a pattern matching:
```
module IntMap = Map.Make (struct
  type t = int
  let compare = compare
end)

let candidates placements =
  match IntMap.bindings (placements) with
  | [(0, hs)] when (List.length hs) >= 2 ->
      (* All VMs in the group run on the hosts to be evacuated *)
      Some hs
  | [(0, hs); (_, [_])] ->
      (* All VMs in the group run on one host *)
      Some hs
  | _ ->
      None

let () =
  let placements =
    [
      (0, ["h0"; "h1"])
    ; (1, ["h2"])
    ]
    |> List.to_seq |> IntMap.of_seq
  in
  match candidates placements with
  | Some hosts ->
      print_endline "To find a host from this list for the group"
  | None ->
      print_endline "No need to place more for the group"
```

Reply:
From the implementation point of view, there is only 1 condition when we need to continue to plan for "no_breach": `resident_hosts_cnt < 2`. The other condition `List.length no_resident_hosts + resident_hosts_cnt > 1` is to shortcut the case when there is only 1 host except the one we are evacuting(we are evacuating from one host to the other one), as we will never got a "no breach" plan, while it will not bring error without shortcut the case, but the logging will looks strange.
To make the code easy to understand, move the shortcut outside of `vm_anti_affinity_no_breach_plan`, as this shortcut also works for `spread evenly` plan.